[AIRFLOW-3314] Changed auto inlets feature to work as described

reubenvanammers opened the pull request:
Currently, the automatic inlets feature is described to be able to skip
operations that produce no outlets. This changes the behaviour of the
auto inlets feature in order so that the presence of non outlet
producing operations are effectively ignored while still receiving all
outlets from upstream tasks.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3314
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Uses a tree search algorithm in order to find upstream outlets to be used as inlets. Provides expected behaviour as described in the previous documentation to receive outlets. 

@bolkedebruin, do you have any input on these changes?
### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Adds unit tests to tests/lineage/test_lineage.py
### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `flake8`

Sorry if I'm breaking any guidelines, this is my first OSS pull request.

exploy commented on airflow/lineage/__init__.py:
```diff
@@ -110,26 +111,32 @@ def wrapper(self, context, *args, **kwargs):
-
-        if len(self._inlets['datasets']) > 0:
-            self.inlets.extend(self._inlets['datasets'])
+        if self._inlets["auto"]:  # Performs a tree traversal, starting with the current task. If outlets
```
Looks like, it should be multiline docsting https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

reubenvanammers replied:
Sure! I wanted to keep the documentation as is since I was writing this PR to match the features of the documentation. On hindsight, since since it does add behaviour not fully covered by the existing docstring, it makes sense to move it into the docstring and clear the behaviour up a little.

bolkedebruin commented on airflow/lineage/__init__.py:
```diff
@@ -110,26 +114,31 @@ def wrapper(self, context, *args, **kwargs):
+                task_id = stack.pop()
+                task = self._dag.task_dict[task_id]
+                visited_task_ids.add(task_id)
+                inlets = self.xcom_pull(
```
This is suboptimal. You will now fire off a database call per `task_id`, which can put quite a strain on the db (although they are indexed lookups). Can you optimize this?

bolkedebruin replied:
I suggest getting all upstream tasks and using a topological sort to have the right order.

reubenvanammers replied:
Hi @bolkedebruin, can you elaborate on your objection? Once the code reaches a node of the DAG and finds outlets, it stops attempting to find further tasks upstream as the upstream tasks  are not placed into the stack. In the case, for example, that the task will have multiple (say, for e.g., 2) upstream tasks, then xcom_pull will be called twice compared to the previous code. However, since xcom_pull fires off a seperate database call (using get_one) for each task_id in the task_ids parameter, I don't think that this should have an appreciable difference in speed. 

Regarding using something like topological sort, I feel like that would lose too much information; in what I regard as the 'expected' behaviour, it requires discrimination between upstream or sibling  nodes, as you want to stop when outlets from a branch are found. As again, since that will do individual database queries, I'm not sure that it will be more efficient.

I could change it so that it doesn't do a database query on the task, but thats a relatively small change. 

Thanks for looking over the PR, and tell me if I have misunderstood something.

reubenvanammers replied:
Hi @bolkedebruin I've squashed the commits, and changed the logic slightly so that it doesn't do a call on the initial task id xcom. Is there a good way to profile the possible slowdown from this commit?

reubenvanammers replied:
Hey @bolkedebruin, I changed the logic a while ago on how the database was called, so it will only do a single db call.  Would you mind having another look?

bolkedebruin reviewed (changes requested):
@reubenvanammers great stuff that you are making some improvements, lineage is still under developed and is inneed of some love :-). Apart from the comment I made can you also squash your commits?

Commit:
[AIRFLOW-3314] Changed auto inlets feature to work as described.

Currently, the automatic inlets feature is described to be able to skip
operations that produce no outlets. This changes the behaviour of the
auto inlets feature in order so that the presence of non outlet
producing operations are effectively ignored while still receiving all
outlets from upstream tasks.
